internal/graph/graph_test.go

```diff
@@ -0,0 +1,156 @@
+// Copyright (c) 2021 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package graph
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+type TestGraph struct {
+	Nodes map[int][]int
+}
+
+func newTestGraph() *TestGraph {
+	return &TestGraph{
+		Nodes: make(map[int][]int),
+	}
+}
+
+func (g TestGraph) Order() int {
+	return len(g.Nodes)
+}
+
+func (g TestGraph) EdgesFrom(u int) []int {
+	return g.Nodes[u]
+}
+
+func TestGraphIsAcyclic(t *testing.T) {
+	testCases := []struct {
+		edges [][]int
+	}{
+		// 1
+		{
+			edges: [][]int{
+				nil,
+			},
+		},
+		// 1 --> 2 --> 3
+		{
+			edges: [][]int{
+				{2},
+				{3},
+				nil,
+			},
+		},
+		// 1 ---> 2 -------> 3
+		// |                 ^
+		// ------------------|
+		{
+			edges: [][]int{
+				{2, 3},
+				{3},
+				nil,
+			},
+		},
+		// 1 --> 2 --> 3    5 --> 6
+		// |           ^    ^
+		// |-----------|    |
+		// ----------> 4 ----
+		{
+			edges: [][]int{
+				{2, 3, 4},
+				{3},
+				nil,
+				{5},
+				{6},
+				nil,
+			},
+		},
+	}
+	for _, tt := range testCases {
+		g := newTestGraph()
+		for i, neighbors := range tt.edges {
+			g.Nodes[i+1] = neighbors
+		}
+		ok, _ := IsAcyclic(g)
+		assert.True(t, ok)
```

Comment:
We could add a message if a cycle was detected:

```suggestion
		ok, cycle := IsAcyclic(g)
		assert.True(t, ok, "expected acyclic, got cycle %v", cycle)
```